api/envoy/config/filter/http/lua/v2/lua.proto

```diff
@@ -18,4 +20,23 @@ message Lua {
   // be properly escaped. YAML configuration may be easier to read since YAML supports multi-line
   // strings so complex scripts can be easily expressed inline in the configuration.
   string inline_code = 1 [(validate.rules).string = {min_bytes: 1}];
+
+  // The Lua source codes that Envoy will execute based on name. Key is Lua source code name, e.g.
+  // ``foo.lua``, this name can be referenced by
+  // :ref:`LuaPerRoute <envoy_api_msg_config.filter.http.lua.v2.LuaPerRoute>`.
+  map<string, api.v2.core.DataSource> source_codes = 2;
+}
+
+message LuaPerRoute {
+  oneof override {
+    option (validate.required) = true;
+
+    // Disable the Lua filter for this particular vhost or route. If disabled is specified in
+    // multiple per-filter-configs, the most specific one will be used.
+    bool disabled = 1 [(validate.rules).bool = {const: true}];
+
+    // The name of a Lua source code defined in
+    // :ref:`Lua.source_codes <envoy_api_field_config.filter.http.lua.v2.Lua.source_codes>`.
+    string name = 2 [(validate.rules).string = {min_bytes: 1}];
```

Comment:
I get that you might want to lookup the code by name for performance/de-dup reasons, but wouldn't you also potentially want to allow the code to be delivered via RDS? I think this would block that? Should this be a oneof that allows either lookup by name or direct inline code?

Reply:
To clarify, "lookup by direct inline code" here is to open to "compile" code on the fly (when the route is being called)?

Reply:
Yeah that's what I meant, or at least have some type of facility to see if we already have a thread local Lua context ready for a per-route code? I'm not sure if we would then need some type of callback to clean this up. I would need to look into all of the ownership stuff in more detail to advise. But I think this is optimally how it would work and we should optimally make sure the API works for that? WDYT?

Reply:
SGTM. I'll update the PR to accommodate that first then.